File: src/lib/idempotency.ts
import crypto from 'crypto'
import { Redis } from '@upstash/redis'

interface MemoryIdempotencyEntry {
  status: 'PROCESSING' | 'COMPLETED'
  response?: any
  expiresAt: number
}

const localMemoryLocks = new Map<string, MemoryIdempotencyEntry>()

// Periodic cleanup of stale in-memory locks every 60s
if (typeof setInterval !== 'undefined') {
  const cleanupTimer = setInterval(() => {
    const now = Date.now()
    for (const [key, entry] of localMemoryLocks.entries()) {
      if (entry.expiresAt <= now) {
        localMemoryLocks.delete(key)
      }
    }
  }, 60000)
  if (cleanupTimer.unref) cleanupTimer.unref()
}

let _redisInstance: Redis | null = null

function getRedis(): Redis | null {
  if (_redisInstance) return _redisInstance
  const url = process.env.UPSTASH_REDIS_REST_URL
  const token = process.env.UPSTASH_REDIS_REST_TOKEN
  if (!url || !token || url === 'placeholder' || token === 'placeholder') return null

  try {
    _redisInstance = new Redis({ url, token })
    return _redisInstance
  } catch {
    return null
  }
}

/**
 * Generate a deterministic cart signature for order idempotency.
 * Combines buyer identifier, sorted cart items, delivery address, and payment method.
 */
export function generateOrderCartSignature(
  buyerId: string,
  items: Array<{ id?: string; productId?: string; quantity?: number; variantId?: string; product?: { id?: string } }>,
  addressId?: string | null,
  paymentMethod?: string | null
): string {
  const normalizedItems = (items || [])
    .map((it) => {
      const pId = it.product?.id || it.productId || it.id || ''
      const qty = it.quantity || 1
      const vId = it.variantId || ''
      return `${pId}:${vId}:${qty}`
    })
    .sort()
    .join('|')

  const raw = `${buyerId.trim()}#${normalizedItems}#${addressId || 'STORE_PICKUP'}#${paymentMethod || 'COD'}`
  return crypto.createHash('sha256').update(raw).digest('hex').slice(0, 32)
}

export interface IdempotencyResult {
  isDuplicate: boolean
  isProcessing?: boolean
  cachedResponse?: any
}

/**
 * Attempt to acquire an atomic lock for order creation.
 * Returns isDuplicate: true if a concurrent or previous identical request was made within ttlSeconds.
 */
export async function acquireIdempotencyLock(
  key: string,
  ttlSeconds = 60
): Promise<IdempotencyResult> {
  const fullKey = `idempotency:order:${key}`
  const now = Date.now()
  const redis = getRedis()

  if (redis) {
    try {
      // 1. Check if key already exists
      const existing = await redis.get<string | { status: string; response: any }>(fullKey)
      if (existing) {
        if (typeof existing === 'string' && existing === 'PROCESSING') {
          return { isDuplicate: true, isProcessing: true }
        }
        if (typeof existing === 'object' && existing.status === 'COMPLETED') {
          return { isDuplicate: true, isProcessing: false, cachedResponse: existing.response }
        }
        return { isDuplicate: true, isProcessing: true }
      }

      // 2. Atomic SET NX with TTL
      const acquired = await redis.set(fullKey, 'PROCESSING', { ex: ttlSeconds, nx: true })
      if (!acquired) {
        return { isDuplicate: true, isProcessing: true }
      }

      return { isDuplicate: false }
    } catch (err) {
      console.warn('Redis idempotency check failed, falling back to in-memory store:', err)
    }
  }

  // Fallback to local memory lock
  const local = localMemoryLocks.get(fullKey)
  if (local && local.expiresAt > now) {
    if (local.status === 'PROCESSING') {
      return { isDuplicate: true, isProcessing: true }
    }
    return { isDuplicate: true, isProcessing: false, cachedResponse: local.response }
  }

  localMemoryLocks.set(fullKey, {
    status: 'PROCESSING',
    expiresAt: now + ttlSeconds * 1000,
  })

  return { isDuplicate: false }
}

/**
 * Cache the successful order creation response for subsequent replay attempts.
 */
export async function saveIdempotencyResponse(
  key: string,
  response: any,
  ttlSeconds = 120
): Promise<void> {
  const fullKey = `idempotency:order:${key}`
  const now = Date.now()
  const payload = { status: 'COMPLETED', response }
  const redis = getRedis()

  if (redis) {
    try {
      await redis.set(fullKey, payload, { ex: ttlSeconds })
    } catch (err) {
      console.warn('Redis saveIdempotencyResponse failed:', err)
    }
  }

  localMemoryLocks.set(fullKey, {
    status: 'COMPLETED',
    response,
    expiresAt: now + ttlSeconds * 1000,
  })
}

/**
 * Release an idempotency lock early (e.g. if order validation failed before database commit).
 */
export async function releaseIdempotencyLock(key: string): Promise<void> {
  const fullKey = `idempotency:order:${key}`
  const redis = getRedis()

  if (redis) {
    try {
      await redis.del(fullKey)
    } catch (err) {
      console.warn('Redis releaseIdempotencyLock failed:', err)
    }
  }

  localMemoryLocks.delete(fullKey)
}

File: fastapi-backend/utils/idempotency.py
import hashlib
import time
import threading
from typing import Dict, Any, Optional, Tuple

# Thread-safe in-memory idempotency locks
_lock = threading.Lock()
_memory_locks: Dict[str, Dict[str, Any]] = {}


def _cleanup_expired():
    now = time.time()
    expired_keys = [k for k, v in _memory_locks.items() if v.get("expires_at", 0) <= now]
    for k in expired_keys:
        _memory_locks.pop(k, None)


def generate_order_cart_signature(
    buyer_id: str,
    items: list,
    address_id: Optional[str] = None,
    payment_method: Optional[str] = None
) -> str:
    """
    Generate a deterministic SHA256 cart signature for order idempotency.
    Combines buyer ID, sorted items (productId:variant:qty), address, and payment method.
    """
    normalized_items = []
    for it in (items or []):
        if not isinstance(it, dict):
            continue
        prod = it.get("product") if isinstance(it.get("product"), dict) else {}
        pid = prod.get("id") or it.get("productId") or it.get("id") or ""
        qty = it.get("quantity") or 1
        var_name = it.get("selectedVariant") or ""
        normalized_items.append(f"{pid}:{var_name}:{qty}")

    normalized_items.sort()
    items_str = "|".join(normalized_items)
    raw = f"{str(buyer_id).strip()}#{items_str}#{address_id or 'STORE_PICKUP'}#{payment_method or 'COD'}"
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:32]


def acquire_idempotency_lock(key: str, ttl_seconds: int = 60) -> Tuple[bool, bool, Optional[dict]]:
    """
    Attempt to acquire an atomic lock for order creation.
    Returns: (is_duplicate: bool, is_processing: bool, cached_response: Optional[dict])
    """
    with _lock:
        _cleanup_expired()
        full_key = f"idempotency:order:{key}"
        now = time.time()

        entry = _memory_locks.get(full_key)
        if entry and entry.get("expires_at", 0) > now:
            if entry.get("status") == "PROCESSING":
                return True, True, None
            return True, False, entry.get("response")

        _memory_locks[full_key] = {
            "status": "PROCESSING",
            "expires_at": now + ttl_seconds
        }
        return False, False, None


def save_idempotency_response(key: str, response: dict, ttl_seconds: int = 120) -> None:
    """
    Cache the successful order creation response for subsequent replay attempts.
    """
    with _lock:
        full_key = f"idempotency:order:{key}"
        now = time.time()
        _memory_locks[full_key] = {
            "status": "COMPLETED",
            "response": response,
            "expires_at": now + ttl_seconds
        }


def release_idempotency_lock(key: str) -> None:
    """
    Release an idempotency lock early if order creation fails before DB commit.
    """
    with _lock:
        full_key = f"idempotency:order:{key}"
        _memory_locks.pop(full_key, None)
